lib/analyze-scope.js

```diff
@@ -1,10 +1,14 @@
 "use strict";
 
 const t = require("@babel/types");
-const escope = require("eslint-scope");
-const Definition = require("eslint-scope/lib/definition").Definition;
-const OriginalPatternVisitor = require("eslint-scope/lib/pattern-visitor");
-const OriginalReferencer = require("eslint-scope/lib/referencer");
+const requireFromESLint = require("./require-from-eslint");
+
+const escope = requireFromESLint("eslint-scope");
+const Definition = requireFromESLint("eslint-scope/lib/definition").Definition;
+const OriginalPatternVisitor = requireFromESLint(
+  "eslint-scope/lib/pattern-visitor"
+);
+const OriginalReferencer = requireFromESLint("eslint-scope/lib/referencer");
```

Comment:
Personally, I don't think that to depend on the internal files of ESLint is a good idea because it's not a public API. ESLint can change the internal structure even in patch releases (ESLint can major update espree/eslint-scope in patch releases if it doesn't affect public behaviors). Instead, it's better if `babel-eslint` depends on own `espree`/`eslint-scope`. Those packages follow semantic versioning.

In my 2 cents.

Reply:
As just a fact, to depend on internal files of ESLint is really fragile. It's possible to be broken on every ESLint release, including patch releases.

As my opinion, it's not good that `babel-eslint` adopts such a fragile implementation. I agree with @kaicataldo. I recommend that `babel-eslint` follows the semantic versioning and update the supported ESLint versions in major versions if needed. The latest `eslint-scope` should be compatible with `eslint@>=5.0.0`.